### PR TITLE
CBL-2973:  Check for the pin-cert with all certs in the chain

### PR DESCRIPTION
**Pull request by jayahariv**
* Check for the pin-cert with all certs in the chain

Unit tests: 
1. **testSelfSignedSSLPinned_SG** : Pins the matched certificate when connecting to a SG configured with a single self-signed certificate. The replicator should be able to connect with the SG without errors.
2. **testSelfSignedSSLFailure_SG** : Pins a non-matched certificate when connecting to a SG configured with a single self-signed certificate. The replicator will not be able to connect with the SG and will be stopped. A TLS error should be reported via the ReplicationChange notification.
3. **testMultipleCertSGWithMatchingCert_SG** : Pins each of the certificate chains when connecting to a SG configured with a chain of three certificates (the chain will contain leaf, intermediate, and root certificate). The replicator should be able to connect with the SG without errors. 
4. **testMultipleCertSGWithNonMatchingCert_SG** : Pins a non-matched certificate when connecting to a SG configured with a chain of three certificates. The replicator will not be able to connect with the SG and will be stopped. A TLS error should be reported via the ReplicationChange notification.
5. **testPinnedCertWithNonTLS_SG** : Pins a certificate but connects to a non-TLS SG. The pinnedServerCertificate will be ignored and the replicator should be able to connect with the SG without errors.

Not added: 
6. **testMultipleCertSGWithExpiredCert_SG** : Pins a matched but expired certificate when connecting to a SG configured with a chain of three certificates. The replicator will not be able to connect with the SG and will be stopped. A TLS error should be reported via the ReplicationChange notification.

### CODE REVIEW

**Comment by jayahariv** [comment]
Jenkins CI Failure is not related
Jira ticket - https://issues.couchbase.com/browse/CBL-3011

File: Objective-C/Tests/ReplicatorTest+Main.m [hunk 1]
```diff
@@ -1763,6 +1772,76 @@ - (void) testPushAndPullExpiredDocument_SG {
...
+    - "SSLCert": "../../../../../Objective-C/Tests/Support/chain-certs/certs.pem",
+    - "SSLKey":  "../../../../../Objective-C/Tests/Support/chain-certs/certs.key",
+ */
+// Replicator with non-matching cert + sg with muliple certs
+- (void) testMultipleCertSGWithNonMatchingCert_SG {
+    id target = [self remoteEndpointWithName: @"itunes" secure: YES];
+    if (!target)
+        return;
+    
+    NSURL* url = [NSURL URLWithString: @"wss://localhost:4986/itunes"];
```
Comment (pasin) [inline Objective-C/Tests/ReplicatorTest+Main.m:1787]: I think there is no need to test with itunes which is big. Testing with scratch (empty) is fine.

File: Objective-C/Tests/ReplicatorTest+Main.m [hunk 2]
```diff
@@ -1763,6 +1772,76 @@ - (void) testPushAndPullExpiredDocument_SG {
...
+    target = [[CBLURLEndpoint alloc] initWithURL: url];
+    
+    self.disableDefaultServerCertPinning = YES;    // without this, SSL handshake will fail
```
Comment (pasin) [inline Objective-C/Tests/ReplicatorTest+Main.m:1790]: I understand that this doesn't pin a non-matched as stated in the test name `testMultipleCertSGWithNonMatchingCert_SG`. Can you pin a non matched or make it into a separate this b/w testing non-match and testing no-pin-cert.

File: Objective-C/Tests/ReplicatorTest+Main.m [hunk 3]
```diff
@@ -1763,6 +1772,76 @@ - (void) testPushAndPullExpiredDocument_SG {
     AssertNil([savedDoc stringForKey: propertyKey]);
 }
 
+#pragma mark - SG (TLS with chain of certs)
+/**
+ - Use the `certs.pem` & `certs.key` in the `LiteCore/Replicator/tests/data/ssl_config.json`
+    - "SSLCert": "../../../../../Objective-C/Tests/Support/chain-certs/certs.pem",
+    - "SSLKey":  "../../../../../Objective-C/Tests/Support/chain-certs/certs.key",
+ */
```
Comment (pasin) [inline Objective-C/Tests/ReplicatorTest+Main.m:1780]: I have updated LiteCore's test to use the cert chain. The new files with be `sg_cert.pem`, `sg_key.pem`, and `cert.pem` (only leaf).

Comment (pasin) [reply Objective-C/Tests/ReplicatorTest+Main.m:1780]: https://github.com/couchbase/couchbase-lite-core/commit/a6401bdcf3580fd71699e4e71b0c8b5abc309942

File: Objective-C/Tests/ReplicatorTest+Main.m [hunk 4]
```diff
@@ -1763,6 +1772,76 @@ - (void) testPushAndPullExpiredDocument_SG {
...
+
+- (SecCertificateRef) certChains: (NSString*)name {
+    NSData* certData = [self dataFromResource: name ofType: @"cer"];
+    SecCertificateRef cert = SecCertificateCreateWithData(NULL, (__bridge CFDataRef)certData);
+    return (SecCertificateRef)CFAutorelease(cert);
+}
+
+// Replicator with single matching cert + sg with muliple certs
+- (void) testMultipleCertSGWithMatchingCert_SG {
+    id target = [self remoteEndpointWithName: @"itunes" secure: YES];
```
Comment (pasin) [inline Objective-C/Tests/ReplicatorTest+Main.m:1803]: Same as above. Just test with scratch.

File: Objective-C/Tests/ReplicatorTest+Main.m [hunk 5]
```diff
@@ -1763,6 +1772,76 @@ - (void) testPushAndPullExpiredDocument_SG {
...
+    
+    NSURL* url = [NSURL URLWithString: @"wss://localhost:4986/itunes"];
+    target = [[CBLURLEndpoint alloc] initWithURL: url];
+    
+    self.disableDefaultServerCertPinning = YES;    // without this, SSL handshake will fail
+    id config = [self configWithTarget: target type: kCBLReplicatorTypePull continuous: NO];
+    [self run: config errorCode: CBLErrorTLSCertUnknownRoot errorDomain: CBLErrorDomain];
+}
+
+- (SecCertificateRef) certChains: (NSString*)name {
```
Comment (pasin) [inline Objective-C/Tests/ReplicatorTest+Main.m:1795]: This method is getting only one cert and cert chain. Probably change to `getCert:`?

Comment (pasin) [reply Objective-C/Tests/ReplicatorTest+Main.m:1795]: Alternatively, you can try something like this. If it works, you just use the PEM string in the test without creating the DER file. I did this in LiteCore test.

```
NSString* pem = "<CERT IN PEM format>"
NSData *certData = [[NSData alloc] initWithBase64Encoding: pem];
SecCertificateRef parsedCertificate = SecCertificateRef cert = SecCertificateCreateWithData(NULL, (__bridge CFDataRef)certData);
```

**Review by pasin — CHANGES_REQUESTED** [review]
I think we should always have PR validation green unless it's really not possible. I recommended to fix PR validation first otherwise we won't be able to use the benefit of the PR validation.

File: Objective-C/Tests/ReplicatorTest+Main.m [hunk 6]
```diff
@@ -1763,6 +1772,110 @@ - (void) testPushAndPullExpiredDocument_SG {
...
 
+#pragma mark - SG (TLS with chain of certs)
+// Replicator with non-matching cert + sg with muliple certs
+- (void) testMultipleCertSGWithNonMatchingCert_SG {
+    id target = [self remoteEndpointWithName: @"scratch" secure: YES];
+    if (!target)
+        return;
+    
+    NSURL* url = [NSURL URLWithString: @"wss://localhost:4994/scratch"];
+    target = [[CBLURLEndpoint alloc] initWithURL: url];
```
Comment (pasin) [inline Objective-C/Tests/ReplicatorTest+Main.m:1783]: Seem like this is not needed as it has been done here:
```
id target = [self remoteEndpointWithName: @"scratch" secure: YES];
```

File: Objective-C/Tests/ReplicatorTest+Main.m [hunk 7]
```diff
@@ -1763,6 +1772,110 @@ - (void) testPushAndPullExpiredDocument_SG {
...
+}
+
+// Replicator with single matching cert + sg with muliple certs
+- (void) testMultipleCertSGWithMatchingCert_SG {
+    id target = [self remoteEndpointWithName: @"scratch" secure: YES];
+    if (!target)
+        return;
+    
+    NSURL* url = [NSURL URLWithString: @"wss://localhost:4994/scratch"];
+    target = [[CBLURLEndpoint alloc] initWithURL: url];
```
Comment (pasin) [inline Objective-C/Tests/ReplicatorTest+Main.m:1803]: Same as above comment.

**Review by pasin — APPROVED** [review]
Two comments, the rest looks good to me.

**Pull request merged**